controller/src/main/java/io/pravega/controller/store/stream/ZkOrderedStore.java

```diff
@@ -83,41 +86,68 @@
-                        storeHelper.createPersistentSequentialZNode(getEntitySequentialPath(scope, stream, latestcollectionNum),
+                .thenCompose(latestCollectionNumber ->
+                        storeHelper.createPersistentSequentialZNode(getEntitySequentialPath(scope, stream, latestCollectionNumber),
                                 entity.getBytes(Charsets.UTF_8))
-                                   .thenCompose(positionPath -> {
-                                       int position = getPositionFromPath(positionPath);
-                                       if (position > rollOverAfter) {
-                                           // if newly created position exceeds rollover limit, we need to delete that entry 
-                                           // and roll over. 
-                                           // 1. delete newly created path
-                                           return storeHelper.deletePath(positionPath, false)
-                                                             // 2. seal latest collection
-                                                             .thenCompose(v -> storeHelper.createZNodeIfNotExist(
-                                                                     getCollectionSealedPath(scope, stream, latestcollectionNum)))
-                                                             // 3. call addEntity recursively
-                                                             .thenCompose(v -> addEntity(scope, stream, entity))
-                                                             // 4. delete empty sealed collection path
-                                                             .thenCompose(orderedPosition -> 
-                                                                     tryDeleteSealedCollection(scope, stream, latestcollectionNum)
-                                                                     .thenApply(v -> orderedPosition));
-
-                                       } else {
-                                           return CompletableFuture.completedFuture(Position.toLong(latestcollectionNum, position));
-                                       }
-                                   }))
+                                .thenCompose(positionPath -> {
+                                    int position = getPositionFromPath(positionPath);
+                                    if (position > rollOverAfter) {
+                                        // if newly created position exceeds rollover limit, we need to delete that entry
+                                        // and roll over.
+                                        try {
+                                            // 1. delete newly created path
+                                            CuratorOp op1 = storeHelper.getClient().transactionOp().delete().forPath(positionPath);
+
+                                            // 2. seal latest collection
```

Comment:
After a further consider, this step is error prone when multi client want to seal latest collection, because all of them try to create the seal node but only one of them would succeed. It seems we cannot swallow `DataExistsException` here. 

Maybe we can exclude it from multi-op and pipe them into

```
seal latest collection -> delete and create in multi-op -> try delete seal node
```

Besides, the original recursive calls check if the new `addEntity` call still exhaust the collection, while in this optimization we don't check so. However, it is almost theoretically only.

Reply:
Updated. Without update pull request description yet.